---
title: "Imputing z-statistics"
subtitle: 'An example analysis'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
    BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Imputing z-statistics}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
---


```{r setup, echo=FALSE, results="hide"}
knitr::opts_chunk$set(
  tidy = FALSE,
  cache = TRUE,
  echo = TRUE,
  dev = c("png", "pdf"),
  package.startup.message = FALSE,
  message = FALSE,
  error = FALSE,
  warning = FALSE
  # collapse = TRUE,
  # comment = "#>",
  # fig.path = ""  # Added this line to the standard setup chunk)
)

options(width = 100)
```	

The `imputez` package uses [GenomicDataStream](https://gabrielhoffman.github.io/GenomicDataStream/) to read genetic data from a reference panel.  Here, simulate z-statstics and use `run_imputez()` to perform the imputation.


### Code
```{r example}
library(imputez)
library(GenomicDataStream)
library(mvtnorm)
library(dplyr)

# VCF file for reference
file <- system.file("extdata", "test.vcf.gz", package = "GenomicDataStream")

# initialize data stream
gds = GenomicDataStream(file, "DS", initialize=TRUE)

# read genotype data from reference
dat = getNextChunk(gds)

# simulate z-statistics with correlation structure
# from the LD of the reference panel
z = c(rmvnorm(1, rep(0, 10), cor(dat$X)))

# Combine z-statistics with variant ID, position, etc
df = dat$info %>%
    mutate(z = z, GWAS_A1 = A1, GWAS_A2 = A2) %>%
    rename(REF_A1 = A1, REF_A2 = A2)

# Given observed z-statistics and 
# GenomicDataStream of reference panel,
# Impute z-statistics from variants missing z-statistics.
# Here drop variant 2, and then impute its z-statistic
# Defaults to run imputezDecorr / decorrelate in the backend
res = run_imputez(df[-2,], gds, 10000, 1000)

# Results of imputed z-statistics
res
```




# Session Info
<details>
```{r session, echo=FALSE}
sessionInfo()
```
</details>




