#' Read plink LD files
#'
#' Read LD files generated by plink into a list of sparseMatrix for each chromosome
#'
#' @param df_files data.frame of file info: chrom, BIM, LD
#'
#' @return list with one entry per chromosome with \code{C.ld} storing the LD correlation matrix and \code{gr_position} storing the variant information
#'
#' @importFrom data.table fread setkey data.table
#' @importFrom GenomicRanges GRanges
#' @importFrom IRanges IRanges
#' @importFrom Matrix sparseMatrix
#' @export
readLDMatrix <- function(df_files) {
  name <- NULL

  if (any(!c("chrom", "BIM", "LD") %in% colnames(df_files))) {
    stop("data.frame must include columns: chrom, BIM, LD")
  }

  # for each chromosome
  r2MatList <- lapply(seq_len(nrow(df_files)), function(i) {
    chrom <- df_files$chrom[i]
    cat("\r", chrom, "   ")

    # read marker position
    df_position <- fread(df_files$BIM[i])
    colnames(df_position) <- c("chrom", "name", "gmap", "position", "allele1", "allele2")

    # read LD
    dfld <- fread(df_files$LD[i], select = c("SNP_A", "SNP_B", "R"))
    dfld <- dfld[!is.nan(dfld$R), ]

    # keep only variants in dfld
    df_position <- df_position[name %in% unique(c(dfld$SNP_A, dfld$SNP_B)), ]

    # create GRanges object for variants in LD matrix
    gr_position <- with(df_position, GRanges(paste0("chr", chrom), IRanges(position, position, name = name), allele1 = allele1, allele2 = allele2))

    # Add self correlation of 1
    df_self <- data.frame(
      SNP_A = df_position$name,
      SNP_B = df_position$name,
      R = 1,
      stringsAsFactors = FALSE
    )
    dfld <- rbind(dfld, data.table(df_self))

    # get index for each marker
    # dfld$idx_A = match(dfld$SNP_A, df_position$name)
    # dfld$idx_B = match(dfld$SNP_B, df_position$name)
    setkey(dfld, "SNP_A", "SNP_B")

    # N = nrow(df_position)

    # # Create sparse matrix
    # C = sparseMatrix( i = dfld$idx_A,
    # 				j = dfld$idx_B,
    # 				x = dfld$R,
    # 				symmetric = TRUE,
    # 				dims = c(N,N),
    # 				dimnames=list(df_position$name, df_position$name) )

    # return LD matrix and variant annotation
    list( # C.ld = C,
      dfld = dfld,
      gr = gr_position
    )
  })

  names(r2MatList) <- as.character(df_files$chrom)

  r2MatList
}

#' Compute fill rate of sparse matrix
#'
#' Compute fill rate of sparse matrix
#'
#' @param M \code{sparseMatrix}
#'
#' @return fraction of entries that are non-zero
#' @export
fillRate <- function(M) {
  if (!is(M, "sparseMatrix")) stop("M must be sparseMatrix")

  if (isSymmetric(M)) {
    rate <- 2 * length(M@x) / nrow(M)^2
  } else {
    rate <- length(M@x) / (nrow(M) * ncol(M))
  }
  rate
}
